src/switch_core_io.c

```diff
@@ -945,6 +985,49 @@ SWITCH_DECLARE(switch_status_t) switch_core_session_read_frame(switch_core_sessi
 	return status;
 }
 
+SWITCH_DECLARE(switch_status_t) switch_core_session_set_fork_read_frame(_In_ switch_core_session_t *session, switch_frame_t *frame)
+{
+	switch_status_t result = SWITCH_STATUS_SUCCESS;
+	switch_mutex_lock(session->fork_read_frame_mutex);
+	if (session->fork_read_frame) {
+		switch_frame_free(&session->fork_read_frame);
+		session->fork_read_frame = NULL;
+	}
+	
+	if (frame) {
+		result = switch_frame_dup(frame, &session->fork_read_frame);
+	}
+	switch_mutex_unlock(session->fork_read_frame_mutex);
+	return result;
+}
+
+SWITCH_DECLARE(switch_status_t) switch_core_session_get_fork_read_frame(_In_ switch_core_session_t *session, switch_frame_t **frame)
+{
+	switch_status_t result = SWITCH_STATUS_FALSE;
+	switch_mutex_lock(session->fork_read_frame_mutex);
+	if (session->fork_read_frame) {
+		result = switch_frame_dup(session->fork_read_frame, frame);
+	}
+	switch_mutex_unlock(session->fork_read_frame_mutex);
+	return result;
+}
+
+SWITCH_DECLARE(switch_status_t) switch_core_session_get_fork_read_frame_data(_In_ switch_core_session_t *session, void *data, switch_size_t* datalen)
+{
+	switch_status_t result = SWITCH_STATUS_FALSE;
+	switch_mutex_lock(session->fork_read_frame_mutex);
+	if (data && session->fork_read_frame && session->fork_read_frame->data && session->fork_read_frame->datalen > 0) {
+		memcpy(data, session->fork_read_frame->data, session->fork_read_frame->datalen);
```

Comment:
The data being pass should be based on the size of SWITCH_RTP_MAX_BUF_LEN which is 16384. fork_read_frame data would only be using SWITCH_RECOMMENDED_BUFFER_SIZE which is 8192. Adding the buffer size for data would add validation for safe copy in case someone will use this in other implementation